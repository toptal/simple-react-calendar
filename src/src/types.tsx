import * as Luxon from 'luxon'

export type CalendarSelectionI = CalendarSelectionRangeI | CalendarSelectionSingleI

export interface CalendarRangeSelectedDate extends CalendarSelectionRangeI {
  inProgress: boolean
}

export interface CalendarSelectionRangeI {
  start: Luxon.DateTime
  end: Luxon.DateTime
}

export type CalendarSelectionSingleI = Luxon.DateTime

export enum CalendarSelectionTypeEnum {
  start = 'start',
  end = 'end',
}

export enum CalendarTypeEnum {
  single = 'single',
  range = 'range',
}

export enum MonthHeaderButtonTypeEnum {
  next = 'next',
  prev = 'previous',
}

export enum NoticeTypeEnum {
  overlap = 'overlapping_with_disabled',
  disabled = 'disabled_day_click',
}
