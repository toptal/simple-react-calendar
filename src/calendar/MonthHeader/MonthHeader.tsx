import React, { FC, SyntheticEvent } from 'react'

import { IMonthHeaderButtonRenderProps, IMonthHeaderDayRenderProps, IMonthHeaderRenderProps } from '../../@types'
import { IGetDaysOfWeek } from '../helper'
import MonthHeaderButton from '../MonthHeaderButton'
import MonthHeaderDay from '../MonthHeaderDay'

export type Props = {
  activeMonth: Date
  blockClassName: string
  customRender?: IMonthHeaderRenderProps
  daysOfWeek: string[]
  getDaysOfWeek: (data: IGetDaysOfWeek) => string[]
  getMonthFormattedName: (day: Date) => string
  handleOnSwitchMonth: (event: SyntheticEvent<HTMLElement>) => void
  isMonthNextEnabled: boolean
  isMonthPrevEnabled: boolean
  renderMonthHeaderButton?: IMonthHeaderButtonRenderProps
  renderMonthHeaderDay?: IMonthHeaderDayRenderProps
  showDaysOfWeek: boolean
  showMonthName: boolean
  showMonthNext: boolean
  showMonthPrev: boolean
  weekStartsOn: 0 | 1 | 2 | 3 | 4 | 5 | 6
}

const MonthHeader: FC<Props> = (props) => {
  const {
    activeMonth,
    blockClassName,
    customRender,
    daysOfWeek,
    getDaysOfWeek,
    getMonthFormattedName,
    handleOnSwitchMonth,
    isMonthNextEnabled,
    isMonthPrevEnabled,
    renderMonthHeaderButton,
    renderMonthHeaderDay,
    showDaysOfWeek,
    showMonthName,
    showMonthNext,
    showMonthPrev,
    weekStartsOn,
  } = props

  if (customRender) {
    return customRender(props)
  }

  if (!showMonthPrev && !showMonthName && !showMonthNext && !showDaysOfWeek) return null

  return (
    <header className={`${blockClassName}-month_header`}>
      <div className={`${blockClassName}-month_header_wrapper`}>
        {showMonthPrev && (
          <MonthHeaderButton
            blockClassName={blockClassName}
            customRender={renderMonthHeaderButton}
            enabled={isMonthPrevEnabled}
            onClick={handleOnSwitchMonth}
            type="prev"
          />
        )}
        {showMonthName && (
          <h3 className={`${blockClassName}-month_header_title`}>{getMonthFormattedName(activeMonth)}</h3>
        )}
        {showMonthNext && (
          <MonthHeaderButton
            blockClassName={blockClassName}
            customRender={renderMonthHeaderButton}
            enabled={isMonthNextEnabled}
            onClick={handleOnSwitchMonth}
            type="next"
          />
        )}
      </div>
      <div className={`${blockClassName}-month_header_days`}>
        {showDaysOfWeek &&
          getDaysOfWeek({dayIndex: weekStartsOn, daysOfWeek}).map((day, index) => (
            <MonthHeaderDay
              blockClassName={blockClassName}
              customRender={renderMonthHeaderDay}
              day={day}
              isWeekend={index >= weekStartsOn + 5}
              key={day}
            />
          ))}
      </div>
    </header>
  )
}

MonthHeader.displayName = 'MonthHeader'

export default MonthHeader
