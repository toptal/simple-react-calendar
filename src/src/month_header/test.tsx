import {shallow} from 'enzyme'
import * as Luxon from 'luxon'
import * as React from 'react'

import MonthHeader, {MonthHeaderProps} from '.'
import {MonthHeaderButtonTypeEnum} from '../types'

const currentTime = Luxon.DateTime.local()

describe('<MonthHeader />', () => {
  let props, wrapper, instance

  beforeEach(() => {
    props = getProps()
    wrapper = shallow(<MonthHeader {...props} />)
    instance = wrapper.instance()
  })

  describe('#handleOnSwitchMonth', () => {
    context('when called with `next`', () => {
      it('calls #onMonthChange', () => {
        instance.handleOnSwitchMonth({currentTarget: {value: MonthHeaderButtonTypeEnum.next}})

        expect(props.onMonthChange).toBeCalledWith(currentTime.plus({month: 1}))
      })
    })

    context('when called with `prev`', () => {
      it('calls #onMonthChange', () => {
        instance.handleOnSwitchMonth({currentTarget: {value: MonthHeaderButtonTypeEnum.prev}})

        expect(props.onMonthChange).toBeCalledWith(currentTime.minus({month: 1}))
      })
    })
  })

  describe('#render', () => {
    it('renders <MonthHeader />', () => {
      expect(wrapper).toMatchSnapshot()
    })

    context('when prop `minDate` is after prop `activeMonth`', () => {
      it('renders <MonthHeader />', () => {
        props = getProps({minDate: currentTime.plus({month: 2})})
        wrapper = shallow(<MonthHeader {...props} />)

        expect(wrapper).toMatchSnapshot()
      })
    })

    context('when prop `maxDate` is before prop `activeMonth`', () => {
      it('renders <MonthHeader />', () => {
        props = getProps({maxDate: currentTime.minus({month: 2})})
        wrapper = shallow(<MonthHeader {...props} />)

        expect(wrapper).toMatchSnapshot()
      })
    })
  })
})

const getProps = (overrides: Partial<MonthHeaderProps> = {}): MonthHeaderProps => ({
  blockClassName: 'example-class-name',
  activeMonth: currentTime,
  onMonthChange: jest.fn(),
  ...overrides,
})
