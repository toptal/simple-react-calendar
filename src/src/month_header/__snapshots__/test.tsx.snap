// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MonthHeader /> #render renders <MonthHeader /> 1`] = `
<div
  className="example-class-name-month_header"
>
  <MonthHeaderButton
    blockClassName="example-class-name"
    isDisabled={false}
    onClick={[Function]}
    type="previous"
  />
  <div
    className="example-class-name-month_header_title"
  >
    January 2000
  </div>
  <MonthHeaderButton
    blockClassName="example-class-name"
    isDisabled={false}
    onClick={[Function]}
    type="next"
  />
</div>
`;

exports[`<MonthHeader /> #render when prop \`maxDate\` is before prop \`activeMonth\` renders <MonthHeader /> 1`] = `
<div
  className="example-class-name-month_header"
>
  <MonthHeaderButton
    blockClassName="example-class-name"
    isDisabled={false}
    onClick={[Function]}
    type="previous"
  />
  <div
    className="example-class-name-month_header_title"
  >
    January 2000
  </div>
  <MonthHeaderButton
    blockClassName="example-class-name"
    isDisabled={true}
    onClick={[Function]}
    type="next"
  />
</div>
`;

exports[`<MonthHeader /> #render when prop \`minDate\` is after prop \`activeMonth\` renders <MonthHeader /> 1`] = `
<div
  className="example-class-name-month_header"
>
  <MonthHeaderButton
    blockClassName="example-class-name"
    isDisabled={true}
    onClick={[Function]}
    type="previous"
  />
  <div
    className="example-class-name-month_header_title"
  >
    January 2000
  </div>
  <MonthHeaderButton
    blockClassName="example-class-name"
    isDisabled={false}
    onClick={[Function]}
    type="next"
  />
</div>
`;
